### PR TITLE
[Storage] Add block_by_version index.

### PR DESCRIPTION
**Pull request by grao1991**
### Description
Add block by version index.
Use it for some block APIs iff sharding is enabled, otherwise still use the old code path.

### Test Plan
Will refactor and add more tests in batch for sharding in a separate PR.

### CODE REVIEW

File: storage/aptosdb/src/db/include/aptosdb_reader.rs [hunk 1]
```diff
@@ -583,30 +583,88 @@ impl DbReader for AptosDB {
...
+                    "Block is not found at height {block_height}, maybe pruned?"
+                ))?
+                .first_version();
             let last_version = self
+                .ledger_db
+                .metadata_db()
+                .get::<BlockInfoSchema>(&(block_height + 1))?
+                .map_or(committed_version, |block_info| {
+                    block_info.first_version() - 1
+                });
```
Comment (msmouse) [inline storage/aptosdb/src/db/include/aptosdb_reader.rs:652]: Could be nice to extract a method?

File: storage/aptosdb/src/db/include/aptosdb_reader.rs [hunk 2]
```diff
@@ -583,30 +583,88 @@ impl DbReader for AptosDB {
...
+                    .lookup_event_after_version(&event_key, first_version)?
+                    .map_or(committed_version, |(v, _, _)| v - 1);
+                return Ok((
+                    first_version,
+                    last_version,
+                    bcs::from_bytes(new_block_event.event_data())?,
+                ));
+            };
+
+            let first_version = self
```
Comment (msmouse) [inline storage/aptosdb/src/db/include/aptosdb_reader.rs:638]: ditto:
```
if !self.skip_index_and_usage {
    way1()
} else {
    way2()
}
```
looks cleaner

**Review by msmouse — APPROVED** [review]

File: storage/aptosdb/src/db/include/aptosdb_reader.rs [hunk 3]
```diff
@@ -583,30 +583,88 @@ impl DbReader for AptosDB {
...
+                .ledger_db
+                .metadata_db()
+                .iter::<BlockByVersionSchema>(ReadOptions::default())?;
+
+            iter.seek_for_prev(&version)?;
+            let (_, block_height) = iter.next().transpose()?.ok_or(anyhow!(
+                "Block is not found at version {version}, maybe pruned?"
+            ))?;
+
+            self.get_block_info_by_height(block_height)
```
[the remark this answers is not in the document]
Comment (grao1991) [reply storage/aptosdb/src/db/include/aptosdb_reader.rs:608]: see my other reply

**Review by areshand — APPROVED** [review]

File: storage/aptosdb/src/db/include/aptosdb_reader.rs (hunk 1 above)
Comment (grao1991) [reply storage/aptosdb/src/db/include/aptosdb_reader.rs]: Done

**Pull request merged**